tap/tapify.py

```diff
@@ -1,106 +1,299 @@
+    "True if you can pass variable/extra kwargs to the class or function (as in **kwargs), else False"
+
+    known_only: bool
+    "If true, ignore extra arguments and only parse known arguments"
+
+
+def _is_pydantic_base_model(obj: Any) -> bool:
+    if inspect.isclass(obj):  # issublcass requires that obj is a class
+        return issubclass(obj, BaseModel)
     else:
-        doc = class_or_function.__doc__
+        return isinstance(obj, BaseModel)
 
-    # Parse docstring
-    docstring = parse(doc)
 
-    # Get the description of each argument in the class init or function
-    param_to_description = {param.arg_name: param.description for param in docstring.params}
+def _tap_data_from_data_model(
+    data_model: Any, func_kwargs: Dict[str, Any], param_to_description: Dict[str, str] = None
+) -> _TapData:
+    """
+    Currently only works when `data_model` is a:
+      - builtin dataclass (class or instance)
+      - Pydantic dataclass (class or instance)
+      - Pydantic BaseModel (class or instance).
 
-    # Create a Tap object with a description from the docstring of the function or class
-    description = "\n".join(filter(None, (docstring.short_description, docstring.long_description)))
-    tap = Tap(description=description, explicit_bool=explicit_bool)
+    The advantage of this function over func:`_tap_data_from_class_or_function` is that descriptions are parsed, b/c
+    this function look at the fields of the data model.
+
+    Note
+    ----
+    Deletes redundant keys from `func_kwargs`
+    """
+    param_to_description = param_to_description or {}
+
+    def arg_data_from_dataclass(name: str, field: dataclasses.Field) -> _ArgData:
+        def is_required(field: dataclasses.Field) -> bool:
+            return field.default is dataclasses.MISSING and field.default_factory is dataclasses.MISSING
+
+        description = param_to_description.get(name, field.metadata.get("description"))
+        return _ArgData(
+            name,
+            field.type,
+            is_required(field),
+            field.default,
+            description,
+        )
+
+    def arg_data_from_pydantic(name: str, field: _PydanticField, annotation: Optional[Type] = None) -> _ArgData:
+        annotation = field.annotation if annotation is None else annotation
+        description = param_to_description.get(name, field.description)
+        return _ArgData(name, annotation, field.is_required(), field.default, description)
+
+    # Determine what type of data model it is and extract fields accordingly
+    if dataclasses.is_dataclass(data_model):
+        name_to_field = {field.name: field for field in dataclasses.fields(data_model)}
+        has_kwargs = False
+        known_only = False
+    elif _is_pydantic_base_model(data_model):
+        name_to_field = data_model.model_fields
+        is_extra_ok = data_model.model_config.get("extra", "ignore") != "forbid"
+        has_kwargs = is_extra_ok
+        known_only = is_extra_ok
```

Comment:
This behavior is more permissive than the current behavior. Currently / in main, both of these are implicitly set to `False`. Lmk if we should keep these as `False` for full backwards compatibility

Run this script in main and then on this branch to see what I mean:

```python
"""
Example:

python demo_data_model.py \
--arg_int 1 \
--arg_list x y z \
--arg_bool
"""
from pydantic import BaseModel, Field
from tap import tapify


class Model(BaseModel):
    """
    My Pydantic Model which contains script args.
    """

    arg_int: int = Field(description="some integer")
    arg_bool: bool = Field(default=True)
    arg_list: list[str] | None = Field(default=None, description="some list of strings")


if __name__ == "__main__":
    model = tapify(Model, not_an_arg=3)
    # In main, you get ValueError: Unknown keyword arguments: {'not_an_arg': 3}
    # In this branch, you don't get a ValueError
    print(model)
```

Reply:
See the Pydantic documentation here: https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.extra

Reply:
Decided to not allow extra args